pkg/provisioners/gcppubsub/dispatcher/dispatcher/reconcile.go

```diff
@@ -0,0 +1,331 @@
+	// running in the background polling the GCP PubSub Subscription.
+
+	if c.DeletionTimestamp != nil {
+		// We use a finalizer to ensure we stop listening on the GCP PubSub Subscriptions.
+		r.stopAllSubscriptions(ctx, c)
+		util.RemoveFinalizer(c, finalizerName)
+		return nil
+	}
+
+	util.AddFinalizer(c, finalizerName)
+
+	err := r.syncSubscriptions(ctx, c)
+	return err
+}
+
+// key creates the first index into reconciler.subscriptions, based on the Channel's name.
+func key(c *eventingv1alpha1.Channel) channelName {
+	return types.NamespacedName{
+		Namespace: c.Namespace,
+		Name:      c.Name,
+	}
+}
+
+// subscriptionKey creates the second index into reconciler.subscriptions, based on the Subscriber's
+// name.
+func subscriptionKey(sub *v1alpha1.ChannelSubscriberSpec) subscriptionName {
+	return types.NamespacedName{
+		Namespace: sub.Ref.Namespace,
+		Name:      sub.Ref.Name,
+	}
+}
+
+// stopAllSubscriptions stops listening to all GCP PubSub Subscriptions for the given Channel.
+func (r *reconciler) stopAllSubscriptions(ctx context.Context, c *eventingv1alpha1.Channel) {
+	r.subscriptionsLock.Lock()
+	defer r.subscriptionsLock.Unlock()
+	r.stopAllSubscriptionsUnderLock(ctx, c)
+}
+
+// stopAllSubscriptionsUnderLock stops listening to all GCP PubSub Subscriptions for the given
+// Channel.
+// Note that it can only be called if reconciler.subscriptionsLock is held.
+func (r *reconciler) stopAllSubscriptionsUnderLock(ctx context.Context, c *eventingv1alpha1.Channel) {
+	channelKey := key(c)
+	if subscribers, present := r.subscriptions[channelKey]; present {
+		for _, subCancel := range subscribers {
+			subCancel()
+		}
+	}
+	delete(r.subscriptions, channelKey)
+}
+
+// syncSubscriptions ensures all subscribers of the Channel have a background Goroutine that is
+// polling the GCP PubSub Subscriptions representing it. It also removes listeners from Subscribers
+// that no longer exist.
+func (r *reconciler) syncSubscriptions(ctx context.Context, c *eventingv1alpha1.Channel) error {
+	r.subscriptionsLock.Lock()
+	defer r.subscriptionsLock.Unlock()
+
+	// We are going to manipulate subscribers, but don't want it written back, so make a throw away
+	// copy.
+	subscribers := c.Spec.Subscribable.DeepCopy()
+	if subscribers == nil {
+		// There are no subscribers.
+		r.stopAllSubscriptionsUnderLock(ctx, c)
+		return nil
+	}
+
+	for _, subscriber := range subscribers.Subscribers {
+		err := r.createSubscriptionUnderLock(loggingWith(ctx, zap.Any("subscriber", subscriber)), c, &subscriber)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Now remove all subscriptions that are no longer present.
+	channelKey := key(c)
+	activeSubscribers := r.subscriptions[channelKey]
+	if len(subscribers.Subscribers) == len(activeSubscribers) {
+		return nil
+	}
+
+	// subsToDelete is logically a set, not a map (values have no meaning).
+	subsToDelete := map[subscriptionName]bool{}
```

Comment:
Minor nit: using `struct{}` as the value saves some allocations. The [set implementations in apimachinery](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/sets/int.go) do this.

FYI those sets are generated; see https://github.com/kubernetes/apimachinery/blob/4a9a8137c0a17bc4594f544987b3f0d48b2e3d3a/pkg/util/sets/types/types.go. If we end up using these sets in multiple places we could do the same.

Reply:
Done.